keps/sig-scheduling/3370-native-podgroup-api/README.md

```diff
@@ -0,0 +1,993 @@
+  Subset string
+}
+```
+
+In this design, a PodGroupStatus is needed to record the latest status. This
+enables users to quickly know what’s going on with this PodGroup; also some
+integrators (like ClusterAutoscaler) can rely on it to make pending PodGroup
+schedulable. For this, we need to add a new PodGroup Controller in
+controller-manager.
+
+```go
+// PodGroupStatus represents the current state of a pod group.
+type PodGroupStatus struct {
+	// Represents the latest available observations of 
+	// a pod group's current state.
+	Conditions []PodGroupCondition
+
+	// Subsets is a list of SubsetStatus which represents 
+	// the current state of a single subset.
+	Subsets []SubsetStatus
+}
+
+// PodGroupCondition describes the state of a pod group 
+// at a certain point.
+type PodGroupCondition struct {
+	// Type of deployment condition.
+	Type PodGroupConditionType
+
+	// Status of the condition, one of True, False, Unknown.
+	Status v1.ConditionStatus
+
+	// The last time this condition was updated.
+	LastUpdateTime metav1.Time
+
+	// Last time the condition transitioned from one status to another.
+	LastTransitionTime metav1.Time
+	
+	// The reason for the condition's last transition.
+	Reason string
+	
+	// A human readable message indicating details about 
+	// the transition.
+	Message string
+}
+
+type PodGroupConditionType string
+
+const (
+	// PodGroupScheduled represents status of the scheduling 
+	// process for this pod group.
+	PodGroupScheduled PodGroupConditionType = "PodGroupScheduled"
+)
+
+// SubsetStatus represents the current state of a subset.
+type SubsetStatus struct {
+	Name string
+
+	// Total number of non-terminated pods targeted by this subset. (
+	// their labels match the selector)
+	Total int32
+
+	// ScheduledAvailable is the number of pods in the subset that can be scheduled
+	// successfully, but the scheduling fails because minMember cannot be met.
+	ScheduledAvailable int32
+
+	// ScheduledAvailable is the number of pods in the subset that can't be 
+	// scheduled successfully.
+	ScheduledUnavailable int32
+
+	// Running is the number of pods targeted by this subset
+	// with a Running phase.
+	Running int32
+
+	// Succeeded is the number of pods targeted by this subset
+	// with a Succeeded phase.
+	Succeeded int32
+
+	// Failed is the number of pods targeted by this subset
+	// with a Failed phase.
+	Failed int32
+}
+```
+
+### User workflow
+1. The user creates a PodGroup object.
+2. The user creates a workload (Deployment, Job, etc.) and associates .spec.podGroup to the previously created PodGroup name.
+3. The user watched the status of PodGroup. Pods associated with this PodGroup are expected to be co-scheduled if resources are adequate.
+4. The user waits for the PodGroup to be complete and then deletes it.
+
+### Implementation
+The implementation details are intentionally left for different schedulers. 
+In terms of Kubernetes default scheduler, we will reveal the details in ano
```

Comment:
More importantly, I don't think we can accept this KEP without an implementation plan for kube-scheduler.

I can think of two high-level implementation approaches:
1. The one that the existing co-scheduling approach is using: Still take a decision for each pod, and hold them in the Wait phase. This needs clarification on what conditions kube-scheduler would set on pods and podgroup status and how cluster-autoscaler would use them.
2. Completely ignore the pods that reference a pod group. The scheduler and autoscaler take a decision on the pod group, not the pods. Only once the podgroup is satisfied and the resources are reserved, the scheduler starts actually scheduling the individual pods. This needs more details around minPods. Also, this implementation is closer to what kueue is doing, so we could potentially use the same API for both. And it's closer to bit.ly/k8s-reservations by @ahg-g. In a high throughput batch cluster, customers are concerned about every object creation. So reducing the number of objects is desirable. And this is my main concern with the k8s-reservations proposal as it is now too.

@MaciekPytel, which high level approach do you think is more appropriate for cluster-autoscaler?

Reply:
Just echoing that we need a detailed implementation plan, this KEP can't move forward without it. I think we need to join forces and try to refine the reservations idea because I feel it will not only address the co-scheduling case, but other cases as well and likely to be more compatible with CA.

Reply:
We do need to fill the implementation section as I commented in https://github.com/kubernetes/enhancements/pull/3371#discussion_r893755404.

For the time being, approach 1 is more feasible, and will qualify the bar of an alpha feature. But I don't see a big difference of 1 vs. 2 in terms of the scheduler implementation. In approach 2, we still need to schedule pods one by one, we cannot take the "reservation" signal as guaranteed, right?

Regarding reservation, I understand this is a de-factor feature of batch scheduling, and also the prerequisite of implementing backfilling. However, I don't interpret it as a _blocker_ of coscheduling. I envision coscheduling a standalone feature, just like other scheduler plugins, and if someday an external controller (like kueue) or core k8s provides the semantics of reservation, it will improve the perf of coscheduling seamlessly/automatically.

Reply:
> But I don't see a big difference of 1 vs. 2 in terms of the scheduler implementation.

The implementation is actually different; in the reservation proposal there are two object APIs, a Reservation object that is 1:1 with pods and a ReservationSet; The scheduler is only aware of the Reservation object and scheduling a reservation is like scheduling a single pod. 

A ReservationSet could have an all-or-nothing scheduling mode implemented by the ReservationSet controller by monitoring the individual Reservations of a ReservationSet and declare the set "scheduled" when the individual reservations are.

> In approach 2, we still need to schedule pods one by one, we cannot take the "reservation" signal as guaranteed, right?

not sure I follow your assumption, but reservations when scheduled are guaranteed just like normal pods getting scheduled now. When scheduling a pod that points to a reservation, it can only schedule if the reservation is already scheduled.

> However, I don't interpret it as a blocker of coscheduling. I envision coscheduling a standalone feature, just like other scheduler plugins, and if someday an external controller (like kueue) or core k8s provides the semantics of reservation, it will improve the perf of coscheduling seamlessly/automatically.

My concerns are twofold: 1) duplicate APIs, or more accurately multiple APIs with overlapping functionality; 2) compatibility with other mature infra built around k8s. 

Regarding concern 1: as you noted the concept of reservation is common in the batch world (it is also useful for other types of workloads), and it can address the co-scheduling case as discussed above. At first glance I don't think the proposed PodGroup API is sufficient to use as part of the reservation APIs, and so we are risking having duplicate/overlapping APIs in the future. 

Regarding concern 2, k8s is now at a different maturity stage than before, major infrastructure is built around it already serving significant customer base. I think the current stage of the co-scheduling plugin doesn't need to be in core k8s, it can continue to exist as is outside core without trouble (it only touches a single component, the scheduler, which already offer a modular way for adding this feature); I believe moving it to core needs more than "For the time being" approach :)

Reply:
> Regarding concern 1: as you noted the concept of reservation is common in the batch world (it is also useful for other types of workloads), and it can address the co-scheduling case as discussed above. At first glance I don't think the proposed PodGroup API is sufficient to use as part of the reservation APIs, and so we are risking having duplicate/overlapping APIs in the future.

Resource reservation is an important concept in the batch world, But gang scheduling is also an import feature in batch world (maybe more common than reservation) . reservation and gang scheduling are two completely separate feature, and it's not good to mix them together; I tend to give users atomic scheduling capabilities. And also PodGroup  API is not only be used in co-scheduling. It a primitive api for batch. It can be used in multiple phases like we want to support preemption between pod group. And also PodGroup Api is factual criteria in the community, so many project use this.

> Regarding concern 2, k8s is now at a different maturity stage than before, major infrastructure is built around it already serving significant customer base. I think the current stage of the co-scheduling plugin doesn't need to be in core k8s, it can continue to exist as is outside core without trouble (it only touches a single component, the scheduler, which already offer a modular way for adding this feature); I believe moving it to core needs more than "For the time being" approach :)

Maybe since 5 years ago, users have been expecting to have the ability to support gang scheduling in kubernetes. Many people have to use other schedulers(like volcano) or build their own scheduler. This creates a lot of extra work for the user. I think if we want to bring the batch users back to native kubernetes. Support for PodGroup and Gang scheduling will be the first step. I don't think pod groups will have any bad consequences for kubernetes. :)

Reply:
>  I don't think pod groups will have any bad consequences for kubernetes.

Again, the major risk is multiple APIs with very similar functionality.

Reply:
> The implementation is actually different; in the reservation proposal there are two object APIs, a Reservation object that is 1:1 with pods and a ReservationSet; The scheduler is only aware of the Reservation object and scheduling a reservation is like scheduling a single pod.

> A ReservationSet could have an all-or-nothing scheduling mode implemented by the ReservationSet controller by monitoring the individual Reservations of a ReservationSet and declare the set "scheduled" when the individual reservations are.

> Again, the major risk is multiple APIs with very similar functionality.

IIUC, Reservation object is 1:1 with pod, right? scheduler will be aware of the reservation object other than ReservationSet. Here you still need something like PodGroup and Gang scheduling to avoid resource deadlocks between two different ReservationSet, right? 

ReservationSet as the top api, ReservationSet controller can create pod group and some reservation objects. I think there is no APIs with similar functionality. They can integrate together.

Reply:
If there is going to be a ReservationSet API, I prefer that kube-scheduler also looks at it, and then there is no need for PodGroup. In other words, PodGroup and ReservationSet would be one object.

Reply:
> If there is going to be a ReservationSet API, I prefer that kube-scheduler also looks at it, and then there is no need for PodGroup.

I disagree PodGroup API can be replaced by or is a subset of ReservationSet. PodGroup literally means: I want this group of pods to be co-scheduled together. In a live cluster, the group of pods may be scheduled immediately **without explicit pre-reserving** resources; or go through a few cycles to be eventually scheduled; or hanging due to resource shortage. It's not _mandatory_ to associate a Preservation with co-scheduling. It's optional.

In terms of Reservation, it doesn't need to involve PodGroup as well. Its semantics is to pre-reserve a chunk of resources and be consumed later. Its scope is much broader than 1:1 mapping with pods. In some batch system's design, it's associated with time. Like, reserve some resources to be used at 10:00 PM, which backfilling can take advantage of - if a workload doesn't have room to run, but can finish before 10:00 PM it's fine for that workload to grab reserved spot.

So my point is: Reservation is wider topic. It needs to be more thoughtfully discussed as it only involves almost all k8s components, and also should leverage some thoughts from traditional batch system.

Reply:
I agree that to implement all-or-nothing over reservations we will need the scheduler to be aware of "ReservationSet", but that doesn't mean it needs to be implemented using the Permit interface. As discussed below, the main difference  is that whatever decision the scheduler makes with regards to reserved capacity will be reflected on the api-server, not only in the scheduler's memory.

>  I don't think pod groups will have any bad consequences for kubernetes even this stage :)

I think it does have bad consequences if there is no autoscaling support from the beginning because CA serves significant user base and it is the component used to provision most cores in the cloud. 

> I disagree PodGroup API can be replaced by or is a subset of ReservationSet. PodGroup literally means: I want this group of pods to be co-scheduled together. In a live cluster, the group of pods may be scheduled immediately without explicit pre-reserving resources; or go through a few cycles to be eventually scheduled; or hanging due to resource shortage. It's not mandatory to associate a Preservation with co-scheduling. It's optional.

One can look at it like this: PodGroup is conceptually a reservation, but the reservation is happening in the scheduler's memory. What makes this problematic in k8s is that the source of truth is an external component, the api-server, which makes it harder for autoscaling to be in sync with what the scheduler is intending to do. If we build all-or-nothing using actual reservations with decisions reflected in the api-server, then we get closer to defining a concrete API between the scheduler and autoscaler. 

>  Mixing with unfinalized design/impl of Reservation would deviate from this KEP's goal and make readers confusing.

I believe there is a legitimate argument that reservations as a wider concept can solve the all-or-nothing use case while being more compatible with autoscaling; so at the minimum we need to discuss this as an alternative.


@MaciekPytel from autoscaling perspective, what are your thoughts on the conceptual difference between PodGroup (as a scheduling primitive only) and Reservations (not necessarily as discussed in bit.ly/k8s-reservations, but as a way to define a more concrete API between the scheduler and autoscaling), which one is more feasible to support via CA and why.

Reply:
> As discussed below, the main difference is that whatever decision the scheduler makes with regards to reserved capacity will be reflected on the api-server, not only in the scheduler's memory.

> from autoscaling perspective, what are your thoughts on the conceptual difference between PodGroup (as a scheduling primitive only) and Reservations (not necessarily as discussed in bit.ly/k8s-reservations, but as a way to define a more concrete API between the scheduler and autoscaling), which one is more feasible to support via CA and why.

Thanks. For this difference, if it's necessary, we can reflect the pod group reserved capacity in the `PodGroup` object on the api-server other than only in the scheduler's memory. The problem we have discussed above can be simplified as How to integrate the scheduler and CA. I think It's not mix the feature for resource reserve and co-scheduling. So we have two choice for this, right?

- once a Pod failed, scheduler would give a failure message (in pod.status.conditions[*]) that CA can read it. I add this choice in the detail design, we can add a reason for scheduled failed call `CoschedulingNotMeetMinMember`.

- Use the `PodGroup` object to integrate the Scheduler and CA, CA can obtain the PodGroup object's spec and status info, CA can decide whether it's possible to resolve the failure by provisioning new machines. For this choice, it's same  as `ReservationSet`.

I am ok for this two choice.  @MaciekPytel  WDYT? I am not sure if it's ok for CA to be aware of `PodGroup` API.

Reply:
> One can look at it like this: PodGroup is conceptually a reservation, but the reservation is happening in the scheduler's memory. What makes this problematic in k8s is that the source of truth is an external component, the api-server, which makes it harder for autoscaling to be in sync with what the scheduler is intending to do.

Essentially, based on this KEP, the implementation a PodGroup's reservation is the same as a single pod's reservation, which doesn't increase the complexity nor needs additional changes to scheduler framework. _If_ we are going to introduce Reservation API, that would change both core logic (assume) and all out-of-tree Reserve/Permit plugins - because based on current Reservation proposal, a Reservation means a chunk of resources are confirmed/secured across all k8s components, so the current in-memory assume/reservation logic is no longer needed.

In other words, if Reservation API gets a consensus, that might be a game-changer - not only PodGroup but also other bits need to be adapted. For now, we shouldn't treat it as a blocker of this KEP as we will still continue the current in-memory reservation mechanism.

Reply:
From CA perspective the challenge is not fetching the PodGroup object, that should be fairly simple.

The problem is that while CA algorithm re-uses scheduler plugins, it works differently than scheduler and it makes stronger assumptions. In particular:
1. CA only runs Filters. The assumption is that Filters are enough to tell if a pod will be able to run on a node provided by CA or not.
2. While point 1 is just how CA happens to be implemented and could in principle be changed, CA uses a greedy binpacking algorithm (I described the details in https://github.com/kubernetes/kubernetes/issues/105802#issuecomment-985776281). 
    * The fact that it's a greedy algorithm means that it assumes that we can make scheduling decisions one-pod-at-a-time.
    * The challenge is that in CA we're solving a different problem than scheduler - we're not just checking if a set of pods can run on a set of nodes, we're trying to find an optimal set of nodes that allows scheduling of all the pods. If at the end of binpacking we find out that one pod in podGroup cannot scheduler, it's not enough that we just "don't schedule" other pods. We have to look for a minimal set of nodes that allows to schedule all the other pods, most likely meaning that we have to start over the entire binpacking logic. We already face major scalability challenges around binpacking, adding arbitrary number of retries (if there are multiple podGroups) doesn't sound like something we can make work in a large cluster.

Because of the above, I think it will be very hard to implement co-scheduling directly in the scheduling cycle, the way that current co-scheduling plugin works. I think an approach that decouples co-scheduling multi-pod check from scheduling of individual pods is much more likely to work with autoscaler. Reservations seem like one possible way of achieving that (assuming minPods check is not part of "scheduling" individual reservation, but something we check separately). Some sort of approach where we "schedule" PodGroup as a whole and not do a check for each pod that is a member of the group (which is how I understand option 2 in https://github.com/kubernetes/enhancements/pull/3371#discussion_r897103506) seems like another feasible approach.

Reply:
@ahg-g I agree with your on [bit.ly/k8s-reservations](https://bit.ly/k8s-reservations) seems like an idea that will work. 

---

@Huang-Wei 
I think this is a different case from nominated pod. In pod preemption scenario the scheduler is effectively saying the pod will be able to schedule shortly and no action is required from CA (a scale-up may be needed later for pods that will be preempted, but those pods will schedule the scale-up normally once they go pending).
With co-scheduling the situation is different - consider an example with 4 pods that requests 1 cpu each and they use co-scheduling, while there are only 3 nodes with 1 cpu each in the cluster. The pods will never be able to schedule without CA intervention and the expected result should clearly be CA adding 1 more node to make the pods become schedulable. 

In order to trigger a correct scale-up, it's not enough for CA to ignore the pods or have a vague information that they're all pending, because of a co-scheduling constraint not being met. CA must be able to run a simulation that will show that adding _just_ 1 node will make all those pending pods schedulable.

---

Re: "BulkFilter" - I'm worried about scalability of this approach and CA. The whole point of CA simulation is to figure out what is the minimum number of nodes that are needed to schedule all the pods. We currently do it by adding nodes one-by-one until there is enough to schedule all the pending pods.
With "BulkFilter" we don't know how many "new nodes" we need to add to schedule all the pods. I'm guessing CA would need to call it in a loop with increasing number of "new nodes" until all the pods are scheduled? The problem is that means we need N separate binpacking runs to find that we need N new nodes. We can try to improve it by doing a binary search or similar, but that is still much more computationally expensive than the current algorithm.

Reply:
@MaciekPytel Re: the similarity of nominated pod, I meant the case that we have enough resources, it's just the quorum not being met (like 3 pods out there, but we require 4), so in CA, it should just ignore the quorum-not-meet pods, and that info is specified by scheduler. Sorry if I didn't make it clear.

Re: the example you listed is exactly the real CA support for co-scheduling, and we do need to let CA be aware that only 1 additional node is needed to schedule all 4 pods.

Re: "BulkFilter", actually that's the cost to support co-scheduling, which IMO is inevitable. Even with reversation, every reservation needs to be simulated and then secured. After reading the pseudo-code in https://github.com/kubernetes/kubernetes/issues/105802#issuecomment-985776281, I think the gap is how to pass  the "(temporary) internal assumed/reserved" (or equivalent) info from scheduler to CA. Say in the above example, 3 out of 4 don't need extra resources, so if CA knows it only needs to simulate the 4th pod, then it would be easier - all other logic can be intact.

Reply:
Reservation can be scheduled incrementally until it is satisfied. In non-autoscaled cluster, it may be a case of waiting for existing pods to finish. With CA, that may mean multiple scale ups, potentially in different node groups.